engine/src/transaction/mod.rs

```diff
@@ -1,107 +1,133 @@
-            Self::Legacy(tx) => &tx.transaction.nonce,
-            Self::AccessList(tx) => &tx.transaction_data.nonce,
-        }
-    }
-
-    pub fn intrinsic_gas(&self, config: &evm::Config) -> Option<u64> {
-        match self {
-            Self::Legacy(tx) => tx.transaction.intrinsic_gas(config),
-            Self::AccessList(tx) => tx.transaction_data.intrinsic_gas(config),
-        }
-    }
+impl TryFrom<&[u8]> for EthTransactionKind {
+    type Error = ParseTransactionError;
 
-    pub fn gas_limit(&self) -> U256 {
-        match self {
-            Self::Legacy(tx) => tx.transaction.gas,
-            Self::AccessList(tx) => tx.transaction_data.gas_limit,
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        if bytes[0] == eip_2930::TYPE_BYTE {
+            let access_list_tx = eip_2930::SignedTransaction2930::decode(&Rlp::new(&bytes[1..]))?;
+            Ok(Self::Eip2930(access_list_tx))
+        } else if bytes[0] <= 0x7f {
+            Err(ParseTransactionError::UnknownTransactionType)
+        } else if bytes[0] == 0xff {
+            Err(ParseTransactionError::ReservedSentinel)
+        } else {
+            let legacy = legacy::LegacyEthSignedTransaction::decode(&Rlp::new(bytes))?;
+            Ok(Self::Legacy(legacy))
         }
     }
+}
 
-    pub fn gas_price(&self) -> U256 {
-        match self {
-            Self::Legacy(tx) => tx.transaction.gas_price,
-            Self::AccessList(tx) => tx.transaction_data.gas_price,
-        }
-    }
+/// A normalized Ethereum transaction which can be created from older
+/// transactions.
+pub struct NormalizedEthTransaction {
+    pub address: Option<Address>,
+    pub chain_id: Option<u64>,
+    pub nonce: U256,
+    pub gas_limit: U256,
+    pub max_priority_fee_per_gas: U256,
+    pub max_fee_per_gas: U256,
+    pub to: Option<Address>,
+    pub value: Wei,
+    pub data: Vec<u8>,
+    pub access_list: Vec<AccessTuple>,
+}
 
-    pub fn destructure(
-        self,
-    ) -> (
-        crate::prelude::Wei,
-        Option<u64>,
-        Vec<u8>,
-        Option<Address>,
-        Vec<AccessTuple>,
-    ) {
-        use crate::prelude::TryInto;
-        match self {
-            Self::Legacy(tx) => {
-                let tx = tx.transaction;
-                (tx.value, tx.gas.try_into().ok(), tx.data, tx.to, Vec::new())
-            }
-            Self::AccessList(tx) => {
-                let tx = tx.transaction_data;
-                (
-                    tx.value,
-                    tx.gas_limit.try_into().ok(),
-                    tx.data,
-                    tx.to,
-                    tx.access_list,
-                )
-            }
+impl From<EthTransactionKind> for NormalizedEthTransaction {
+    fn from(kind: EthTransactionKind) -> Self {
+        use EthTransactionKind::*;
+        match kind {
+            Legacy(tx) => Self {
+                address: tx.sender(),
+                chain_id: None, // This should be fine as it'll fail regardless.
```

Comment:
This looks wrong. Why not properly derive the `chain_id` from the `v` field?

Reply:
Yeah that should be updated.